### PR TITLE
Add perm_cast_simple

### LINKED ISSUE #31: Randomly insert casts

### PR DESCRIPTION
**Pull request by Synray**
Resolves #31

### CODE REVIEW

File: src/randomizer.py [hunk 1]
```diff
@@ -905,6 +905,52 @@ def visitor(expr: Expression) -> None:
...
+    expr = random.choice(cands)
+    type: SimpleType = decayed_expr_type(expr, typemap)
+    base_type = resolve_typedefs(type, typemap)
+
+    if not isinstance(base_type, ca.TypeDecl):
+        return False
+    if not isinstance(base_type.type, ca.IdentifierType):
+        return False
+    if all(x not in base_type.type.names for x in ['int', 'char', 'long', 'short', 'signed', 'unsigned', 'float', 'double']):
+        return False
```
Comment (simonlindholm) [inline src/randomizer.py]: this pattern occurs in some other place as well, can we make a helper for it?

File: src/randomizer.py [hunk 2]
```diff
@@ -905,6 +905,52 @@ def visitor(expr: Expression) -> None:
...
+
+    integral_type: List[List[str]] = [['int'], ['char'], ['long'], ['short'], ['long', 'long']]
+    floating_type: List[List[str]] = [['float'], ['double'], ['long', 'double']]
```
Comment (simonlindholm) [inline src/randomizer.py]: let's remove long double, IDO doesn't support it, and it seems like it should be really easy to spot by hand.

File: src/randomizer.py [hunk 3]
```diff
@@ -905,6 +905,52 @@ def visitor(expr: Expression) -> None:
...
+    new_type: List[str]
+    if random.choice([True, False]):
+        # Cast to integral type, sometimes unsigned
+        sign: List[str] = random.choice([[], ['unsigned']])
+        new_type = sign + random.choice(integral_type)
+    else:
+        # Cast to floating point type
+        new_type = random.choice(floating_type)
+
+    visit_replace(fn.body, lambda n, _: ca.Cast(ca.Typename(name=None,quals=[], type=ca.TypeDecl(None, [], ca.IdentifierType(new_type))), expr) if n is expr else None)
```
Comment (simonlindholm) [inline src/randomizer.py]: pretty ugly one-liner; break out into a separate function and/or format with black.

File: src/randomizer.py [hunk 4]
```diff
@@ -905,6 +905,52 @@ def visitor(expr: Expression) -> None:
...
+    # Find a random expression
+    def rec(block: Block) -> None:
+        for stmt in ast_util.get_block_stmts(block, False):
+            ast_util.for_nested_blocks(stmt, rec)
+            def visitor(expr: Expression) -> None:
+                if not region.contains_node(expr):
+                    return
+                cands.append(expr)
+            replace_subexprs(stmt, visitor)
+    rec(fn.body)
```
Comment (simonlindholm) [inline src/randomizer.py]: would be nice to make this a helper function

**Comment by Synray** [comment]
Also halved mask and cast_simple's probabilities to 10, they sometimes generate identical asm.

**Pull request merged**